File: tests/observability.py
from __future__ import annotations

import json
from dataclasses import asdict, is_dataclass
from pathlib import Path
from typing import Any


ARTIFACT_DIR = Path("tests/_artifacts")


def write_artifact(filename: str, content: str) -> Path:
    ARTIFACT_DIR.mkdir(parents=True, exist_ok=True)
    path = ARTIFACT_DIR / filename
    path.write_text(content)
    return path


def write_json_artifact(filename: str, data: Any) -> Path:
    if is_dataclass(data):
        payload = asdict(data)
    else:
        payload = data
    text = json.dumps(payload, indent=2, ensure_ascii=True)
    return write_artifact(filename, text)

File: frontend/tests/observability.ts
import fs from "node:fs";
import path from "node:path";

const ARTIFACT_DIR = path.resolve(process.cwd(), "tests/_artifacts");

export function writeArtifact(filename: string, content: string): string {
  fs.mkdirSync(ARTIFACT_DIR, { recursive: true });
  const target = path.join(ARTIFACT_DIR, filename);
  fs.writeFileSync(target, content, "utf-8");
  return target;
}

export function writeJsonArtifact(filename: string, data: unknown): string {
  const content = JSON.stringify(data, null, 2);
  return writeArtifact(filename, content);
}
